Make a `Data` object part of the `BaseModel.__init__`

davidorme opened the pull request:
# Description

This PR:

1. Makes a `Data` object a core part of model creation, by adding it to the `__init__` method for the `BaseModel` ABC. All subclasses now have a way to connect through to a shared `Data`instance.
2. Adds the `required_init_vars` property to `BaseModel` and adds running the new `check_required_init_vars` to the `BaseModel.__init__` method. Model subclasses can now define what variables (and on what axes) they need in Data to successfully `__init__` via that property. (Discussed #97)
3. Makes the two class properties (`model_name` and `required_init_vars`) part of the abstract BaseModel API, rather than just class attributes.
4. Updates `__init_subclass__`
5. Updates `test_model.py` to add new tests for.
6. Updates module docstring.
7. Updates `main.py`, `test_main.py` and existing models to match new signatures.

Fixes #97

Some notes:

* I wanted to use a parameterised test for `__init_subclass_` because there are multiple new subclass failure modes. That is awkward because the new class has to be defined within the test. I've used `exec()` to run the different subclass attempts within the model. That feels a bit clumsy but I can't see a better way and I don't think it is actively harmful.
* This PR has revealed that we need a new `Grid.from_config` function but I'll leave that for another PR.
* The `test_main.py` tests are using the `SoilModel` as a demo - that is a constantly moving target as that model develops so we should probably just define a new test BaseModel subclass in here for use in testing `main.py`.


## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

davidorme commented:
Just fixing the conflicts arising from the recently merged #174

vgro commented on virtual_rainforest/core/model.py:
```diff
@@ -1,61 +1,69 @@
-:class:`~virtual_rainforest.core.model.BaseModel` can been seen in the :mod:`soil.model`
-module.
+"""The :mod:`~virtual_rainforest.core.model` module defines the high level API for the
+different models within the Virtual Rainforest. The module creates the
```
for consistency, should we call this `core_model`?

jacobcook1995 replied:
I think it's probably better to call it something like `model_api`. The reason I didn't update the name here to match the other model naming scheme is that it isn't really a model that would be used in the same way as other models, which giving it a consistent name could implicitly suggest that it was

davidorme replied:
I think giving it a less ambiguous name would be good. I'd lean towards `base_model` - that seems to me clearer that it is the base for real models (and the object is called BaseModel!). `base_abc_model` would be even clearer but seems a bit over the top.

Perhaps make this a separate PR?

vgro replied:
Side note: once we have an agreement on this, could you please update the API links to the updated model names for soil and abiotic? We missed that when updating the model names.

jacobcook1995 replied:
Yeah I would be happy with `base_model.py`

vgro reviewed (approved):
As far as I can tell, this all makes sense :-)

jacobcook1995 commented on tests/test_models.py:
```diff
@@ -5,339 +5,295 @@
+    """Creates a simple data instance for use in testing."""
+    from xarray import DataArray
 
-    # In cases where it passes then checks that the object has the right properties
```
Is this section being removed because the functionality is tested when inheriting classes are tested for this?

davidorme replied:
Yup - I'm running this testing on an example minimal real class rather than making it possible to create an instance of BaseModel itself. The `test_check_failure_on_missing_methods` tackles this specifically. It would also catch undefined `model_name` and `required_init_vars` but those are checked in more detail in `__init_subclass__` so cannot  be missing when an instance is created.

jacobcook1995 reviewed (approved):
This all seems sensible to me. Added a few queries, but nothing significant enough to prevent merging if others are happy for it to be merged

davidorme commented on virtual_rainforest/core/model.py:
```diff
@@ -118,34 +163,58 @@ def cleanup(self) -> None:
+
+        if not isinstance(cls.model_name, str):
+            excep = TypeError(f"Property model_name in {cls.__name__} is not a string")
+            LOGGER.error(excep)
```
Good point. Same for the other ones here too - they will all occur when importing `virtual_rainforest`.

jacobcook1995 commented:
Actually while I remember, once the changes to setting up models are set in stone, would you be able to update `docs/development/defining_new_models.md` to reflect the changes?

davidorme commented:
@jacobcook1995 I'll do that in a mini PR, once #176 and #173 are in `develop`, otherwise I'll be documenting across branches 😖

jacobcook1995 commented:
Yeah that makes sense

dalonsoa commented on virtual_rainforest/core/base_model.py:
```diff
@@ -119,34 +163,58 @@ def cleanup(self) -> None:
+
+        # Check that model_name exists and is a string - if it is not implemented in the
+        # subclass then the object will be of type property
+        if isinstance(cls.model_name, property):
+            excep = NotImplementedError(
+                f"Property model_name is not implemented in {cls.__name__}"
+            )
```
Given that `model_name` is an `abstractmethod`, shouldn't be just impossible to create a subclass without overwriting it? That's the whole point of `abstractmethods`, right?

davidorme replied:
I got hung up on this too! The distinction is (hat tip, @jacobcook1995) that the automatic testing for missing `abstractmethod` objects only happens when you try and create an instance of the class. When `__init_subclass__` runs, it is just dealing with the `cls` object. In this function, we want to check the name is valid, but do need to gatekeep first for undefined `model_name`. 

Arguably, we could leave out the `required_init_vars` - we don't currently test the contents - and then it would get flagged at instantiation if missing. I think testing the contents is probably valuable though.

dalonsoa replied:
Interesting. Indeed, I didn't know abstract classes worked that way. Yes, I agree it is useful to check during the definition of the subclasses.

davidorme replied:
Is this kind of structure using `assert` ok for checking the contents of `required_init_vars`. I've not really used `assert` like this before but it seems a compact way to raise exceptions without having to nest a bunch of `if` statements to catch specific failure modes:

```python
        # Check the structure
        try:
            assert isinstance(cls.required_init_vars, tuple)
            for vname, axes in cls.required_init_vars:
                assert isinstance(vname, str)
                assert isinstance(axes, tuple)
                assert all([isinstance(a, str) for a in axes])
        except (AssertionError, ValueError):
            raise ValueError(
                f"Property required_init_vars has the wrong structure in {cls.__name__}"
            )
```

dalonsoa replied:
There is nothing conceptually wrong with using `assert` I think, but `assert` checks are disabled when running python with the `-O` flag. From `python -h`:

```bash
-O     : remove assert and __debug__-dependent statements;
```

I'm not sure if you intend to run the code in optimised mode, but just in case I would not use `assert` to control the flow of the code.

davidorme replied:
Ah. Damn. Yes - we don't want error checking to disappear in optimised mode. More long winded approach it is then.

dalonsoa commented on virtual_rainforest/core/base_model.py:
```diff
@@ -119,34 +163,58 @@ def cleanup(self) -> None:
+            raise excep
+
+        # Check that required_init_vars is set - TODO - not testing structure here
+        if isinstance(cls.required_init_vars, property):
+            excep = NotImplementedError(
+                f"Property required_init_vars is not implemented in {cls.__name__}"
+            )
```
Ans the same here.

dalonsoa commented on tests/test_main.py:
```diff
@@ -24,10 +24,20 @@
 from .conftest import log_check
 
 
+@pytest.fixture
+def data_instance():
+    """Creates an emmpty data instance."""
+    from virtual_rainforest.core.data import Data
+    from virtual_rainforest.core.grid import Grid
+
+    grid = Grid()
+    return Data(grid)
```
Given that this fixture is identical to the one in `test_soil_model`, I would add it to a `conftest.py` file in the tests root directory.

dalonsoa reviewed (approved):
All looks good. I think it is the right move.

I just made a couple of comments. It might be I don't understand how abstract classes work, in particular in reference to properties.

alexdewar commented on tests/test_main.py:
```diff
@@ -24,10 +24,20 @@
 
+@pytest.fixture
+def data_instance():
+    """Creates an emmpty data instance."""
```
Spelling

alexdewar commented on virtual_rainforest/models/abiotic/abiotic_model.py:
```diff
@@ -66,22 +69,23 @@ def __init__(
         invalid rather than returning an initialised model instance None is returned.
 
         Args:
+            data: A :class:`~virtual_rainforest.core.data.Data` instance.
```
Minor nit: the user will already be able to see `data`'s type, so there's no point in repeating it. How about describing it as "loaded user data" or something?

alexdewar commented on virtual_rainforest/models/soil/soil_model.py:
```diff
@@ -42,9 +43,11 @@ class SoilModel(BaseModel):
 
     model_name = "soil"
     """The model name for use in registering the model and logging."""
+    required_init_vars = []
```
How about using a type annotation here so that `mypy` knows what *kind* of list it will be?

davidorme replied:
Does it not _have_ to follow the type of the abstract property in BaseModel? `mypy` certainly throws an error when it doesn't.

Having said that - it seems that we _have_ to manually add docstrings for the subclasses or these properties don't get documented. I think we probably could do something clever and make them inheritable (probably in `__init_subclass__` but they probably should be subclass specific.

alexdewar reviewed (approved):
I've made some minor comments, but it all looks good to me.

davidorme commented:
OK - I have now updated this to check `required_init_vars` has the right structure and uses known axes when the `__init_subclass` runs. That means that unknown core axes from bad model setup don't get flagged when users are trying to use the model class - the whole thing falls over when `vr` is imported. That seems correct for a fundamental programming error.

I've also updated the docstrings for existing models and the API stubs in source to better document the new data argument and the subclass attributes.

davidorme commented:
@jacobcook1995 I've updated the new models document to match - can you have a look?

dalonsoa commented on tests/models/abiotic/test_abiotic_model.py:
```diff
@@ -11,6 +11,16 @@
 from virtual_rainforest.models.abiotic.abiotic_model import AbioticModel
 
 
+@pytest.fixture
+def data_instance():
+    """Simple data instance to use in model init."""
+    from virtual_rainforest.core.data import Data
+    from virtual_rainforest.core.grid import Grid
+
+    grid = Grid()
+    return Data(grid)
+
+
```
```suggestion
```
As this is now in `conftest.py`, it can be removed from here.

dalonsoa commented on virtual_rainforest/core/base_model.py:
```diff
@@ -119,34 +177,96 @@ def cleanup(self) -> None:
     @classmethod
     def __init_subclass__(cls) -> None:
-        """Method that adds new model classes to the model registry.
+        """Initialise subclasses deriving from BaseModel.
```
The functionality of this method is totally ok, but I would split it into smaller functions for checking the individual attributes, leaving this as a top level method that just call those internally. That will make understanding what's going here on and testing easier.

davidorme replied:
This was _much_ more irritating to do than anticipated. `mypy` basically complained about _everything_ once the components of the type checking (which is essentially the whole of the function) were isolated in separate functions. And at the end, we now have `# type: ignore`s because `mypy` in `__init_subclass` is no longer aware that the properties have been checked and have the correct structure.

Sometimes I _hate_ `mypy`.

davidorme replied:
I mean, I agree that the end point is cleaner, but.... gaaaaaah!

dalonsoa commented on virtual_rainforest/core/base_model.py:
```diff
@@ -160,7 +280,57 @@ def __repr__(self) -> str:
-            return "A base model instance"
+        return f"A {self.model_name} model instance"
+
+    def check_required_init_vars(self) -> None:
```
Just to be sure I get this: the checks in `init_subclass` are done when setting up the class - i.e. when importing `vr` and these ones are checked when running the simulation, at the config stage, right? 

I would rename this method because it seems a bit confusing checking the `required_init_vars` in two places considering that in one case you are checking the class and in another the contents of data.

davidorme replied:
That's correct on the functionality - and I agree that the naming is ambiguous.

dalonsoa reviewed (changes requested):
I've made a couple of suggestions, maybe to improve clarity of the code. The functionality itself is totally fine.

jacobcook1995 commented on tests/core/test_base_model.py:
```diff
@@ -1,95 +1,350 @@
 
-This module tests the functionality of base_model.py, which defines the basic model API
-that specific models (e.g. `soil_model.py`) utilise.
+This module tests the functionality of model.py, as well as other bits of code that
```
It doesn't actually test other bits of code anymore as this has been moved to separate scripts

jacobcook1995 reviewed (approved):
The doc changes are good! Have added a few minor comments

davidorme commented on docs/source/development/defining_new_models.md:
```diff
@@ -93,23 +167,116 @@ def __init__(
+        "update_interval": {
+            "type": "string",
+            "default": "",
+            "title": "The update_interval Schema",
```
Those fields are straight out of the JSONSchema app link shown above. I might add a note to say that we don't need all of them!

alexdewar reviewed (approved):
LGTM!

dalonsoa reviewed (approved):
Well, `mypy` can be annoying, but it is easy to silence it and, in this case, it is justified and clear, so ignoring it is not dangerous.
